File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <queue>

using namespace std;


vector<int> visited(1000002, -1);
int target =0;
int add = 0;

void search(int num)
{
    queue <pair<int, int>> q;
    
    q.push({num, 0});
    
    visited[num] = 0;
    
    while(!q.empty()) 
    {
        auto& item = q.front();
        q.pop();
        
        int cur = item.first;
        int depth = item.second;
        
        if(cur == target) continue;
        
        //다음 노드 넣기
        vector<int> v;
        v.push_back(cur*3);
        v.push_back(cur*2);
        v.push_back(cur + add);
        
        for(int& tmp : v)
        {
            if(tmp <= target)
            {
                if(visited[tmp] == -1) {
                    q.push({tmp, depth +1});
                    visited[tmp] = depth +1;
                }
            }
        }
    }
}

int solution(int x, int y, int n) {
    int answer = 0;
    
    target = y;
    add= n;
    
    search(x);
    
    answer = visited[y];
    
    return answer;
}

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.js
function solution(x, y, n) {
    var answer = -1;
    
    let queue = [[y, 0]];
    
    while(queue.length !== 0) {
        const item = queue[0];
        const num = item[0];
        const depth = item[1];
        queue.shift();
        
        if(x === num) return depth;
        
        if(num %3 === 0) {
            const tmp = num / 3;
            if(tmp > 0) queue.push([tmp, depth+1]);
        }
        
        if(num %2 === 0) {
            const tmp = num / 2;
            if(tmp > 0) queue.push([tmp, depth+1]);
        }
        
        if(num - n > 0) {
            queue.push([num-n, depth+1]);
        }
        
    }
    
    
    return answer;
}